### ISSUE TITLE
Unexpected results of search

### ISSUE THREAD

**Issue by gdubicki** [wwkimball/yamlpath#206]
## Operating System

1. Name/Distribution: macOS 
2. Version: Monterey 12.6.1 (21G217)

## Version of Python and packages in use at the time of the issue.

1. [Distribution](https://wiki.python.org/moin/PythonDistributions): CPython from homebrew
2. Python Version: 3.11.0
3. Version of yamlpath installed: 3.6.9
4. Version of ruamel.yaml installed: 0.17.21

## Minimum sample of YAML (or compatible) data necessary to trigger the issue

```yaml
projects_and_groups:

  "*":
    merge_requests_approval_rules:
      standard:
        approvals_required: 1
        name: "All eligible users"

  "foo/bar":
    merge_requests_approval_rules:
      dev-uat:
        groups:
          - some_group
        users:
          - a_user
```

## Complete steps to reproduce the issue when triggered via:

YAML Path: `"**.merge_requests_approval_rules.*[.=users]"`

## Expected Outcome

1 result, the array at "projects_and_groups."foo/bar".merge_requests_approval_rules.dev-uat.users".

## Actual Outcome

Exception: `Cannot add PathSegmentTypes.MATCH_ALL subreference to dictionaries at 'None' in '**.merge_requests_approval_rules.*[.=users]'.`

**Comment by gdubicki** [comment]
The code in question is in this PR https://github.com/gitlabform/gitlabform/pull/467/files?diff=split&w=1, in the `gitlabform/configuration/transform.py` file, the `UserTransformer` class, `_do_transform` method.

The test that shows how it fails is in the same PR, in the `tests/unit/configuration/transform/test_user_and_group_transformers.py` file.

<br/>

What I find strange here is that without the `"*":` dict key:

```yaml
projects_and_groups:

  # "*":
  #   merge_requests_approval_rules:
  #     standard:
  #       approvals_required: 1
  #       name: "All eligible users"

  "foo/bar":
    merge_requests_approval_rules:
      dev-uat:
        groups:
          - some_group
        users:
          - a_user
```
...my code works like expected - I am getting a single, correct result.

<br/>

What is even more strange to me is that if I comment out _part_ of the `"*":` dict key like this:

```yaml
projects_and_groups:

  "*":
    merge_requests_approval_rules:
      # standard:
      #   approvals_required: 1
      #   name: "All eligible users"

  "foo/bar":
    merge_requests_approval_rules:
      dev-uat:
        groups:
          - some_group
        users:
          - a_user
```
...I am getting 2 results, but one of them is `None`. I hope that this screenshot from IntelliJ provides enough data about this case:
<img width="1438" alt="Screenshot 2023-01-01 at 11 59 48" src="https://user-images.githubusercontent.com/566632/210168363-ea8155a5-077a-44ce-8482-d305b2b24f06.png">

**Comment by gdubicki** [comment]
Please let me know if you need any more info, additional context etc., @wwkimball! I didn't want to bother you with my issue but I really got stuck and I don't know if I am using your library wrong or if it's a bug.

**Comment by wwkimball** [comment]
Thank you for this report; I'll take a look.

**Comment by wwkimball** [comment]
@gdubicki Is there by chance any more to the data or any special configuration of the library?  When I run your data through the `yaml-get` command, I am not getting any error:

```
$ python --version
Python 3.11.0

$ yaml-get --version
yaml-get 3.6.9

$ cat 206.yaml 
projects_and_groups:

  "*":
    merge_requests_approval_rules:
      standard:
        approvals_required: 1
        name: "All eligible users"

  "foo/bar":
    merge_requests_approval_rules:
      dev-uat:
        groups:
          - some_group
        users:
          - a_user

$ yaml-get --query='**.merge_requests_approval_rules.*[.=users]' 206.yaml 
["a_user"]

$ echo $?
0
```

**Comment by wwkimball** [comment]
I'm guilty of only reading your initial post.  I see now that you've added a lot more context.  I'm trying to sort through it all, but there's quite a lot in the indicated PR to parse through.  Can you provide a minimum code example which reproduces the error?

As it sits right now, I can't reproduce the error.  That leads me to think the issue is somewhere in your code or there's something funky in the rest of the YAML document that isn't visible in the sample you'd provided here.  I'm happy to be wrong; if there's a bug in yamlpath, I very much want to work it out!

**Comment by wwkimball** [comment]
Never-mind.  I couldn't leave this alone and ran some experiments until I was finally able to reproduce this issue.  It evidently occurs only when the query permits arbitrary node creation, which adds all nodes defined in the YAML Path when they don't already exist in the data.  This is the default mode for the Processor's [get_nodes](https://github.com/wwkimball/yamlpath/wiki/processor.py#public-instance-methods) method.  This is useful while _setting_ data, when the DOM may lack some or all of the parentage of the target node.  But when _querying_ data, you should turn this off by setting `mustexist=True`, [as illustrated in the documentation](https://github.com/wwkimball/yamlpath/wiki/Python-Library#getting-data-from-the-document).

In your specific use-case, I've added a comment to your PR at https://github.com/gitlabform/gitlabform/pull/467/files?diff=split&w=1#r1064063152 where you can fix this issue.  I'll think over what I can do to protect against unintentional wildcard ancestry generation, or perhaps a clearer error message when `**` is attempted during a query operation.

Please let me know if I can help any further.

**Comment by gdubicki** [comment]
Big, big thank you for resolving this completely! <3

>  This is useful when your code has a default value it wants when the target path does not exist and you'd like it to be put in the document as part of the query.

To be honest I don't understand that use case... But maybe it's just me if no one else complained.

> I'll think over what I can do to protect against unintentional wildcard ancestry generation, or perhaps a clearer error message when ** is attempted during a query operation while mustexist is False.

How about logging a warning? Does such an operation even make sense? 🤔

**Comment by wwkimball** [comment]
Taking a look at this behavior reminds me that mutating data as a result of a read (get) operation is abhorrent.  Outside of just hammering out code faster than I could think deeply about every decision, I don't know why I ever wrote the get method to behave that way by default.  It makes me wonder what the ramifications might be of reversing the default...  it'd be a major API change that would change default behavior.  A change to default behavior would adversely affect users who are already relying on the present behavior.

It's easy to imagine a scenario where "add missing document structure and write this default value to the new node" would be useful.  It takes more code to "try get; on fail get, try write default" than "try get (knowing that the default write will take care of itself)".  It's a kind of syntax-sugar, in a sense.  When you don't want this convenience, set `mustexist=True`.  Reversing this, we'd be informing users that "if you want the convenience of automatic DOM generation for missing keys, set `mustexist=False`".  It's an inversion of the same thought.

I suspect no one else has come forward with this as an issue simply because they copy-paste the sample code and carry on.

I'm mulling over the possibility of allowing arbitrary node creation when `**` is used.  On one hand, it could be convenient to mass-generate a document's structure with some default value(s).  On the other, this could be extremely dangerous, setting values potentially all over a document should the user not sufficiently predict everywhere `**` will go.  The potential for harm may outweigh the conceptual benefit.

I welcome your thoughts on this!

**Comment by AndydeCleyre** [comment]
I'll say that I was also surprised by this behavior and had to fix my code: https://github.com/AndydeCleyre/nestedtextto/commit/5eed778bbf31a8e23fa4767037b4171b73f48b79

I'd expect terminology closer to dict's setdefault.

**Comment by wwkimball** [comment]
This concern turns up where the `*` segment appears.  As an experiment, I switched the default behavior but that triggers about a dozen unit-test failures.  I could certainly fix the unit tests, but that again brings up the specter of end-users who may have already written code which depends on this (arguably bad) default behavior.

In better news, I also experimented with simply excluding `**` and `*` from the code body that runs when `mustexist=False` and the YAML Path has so far matched nothing, largely because I couldn't think of a use-case where people would actively want to return non-existent nodes.  That doesn't cause any failures and so far, I haven't come up with a use/test-case where I want non-existent nodes to be created during a get operation.

The suggestion from @AndydeCleyre is interesting but not simple because the Processor returns nodes, not values.  In order to return a node for a non-existent path, a node must be created where the path points, with full ancestry.  To wit, that's what the code is already attempting to do.  It just gets lost when the wildcard is used because there's no root for the ancestry when every node in the real and constructed DOM could be the root.

I'll keep thinking this over.  Thanks for any and all suggestions!!

**Comment by wwkimball** [comment]
Here's my CHANGELOG proposal for a fix:

```
* Processor.get_nodes would emit a nonobvious error message when mustexist is
  left at its default value (False) and yaml_path contained ** or * segments.
  Now, these segment types are excluded from generating "missing" nodes; only
  nodes which exist can be matched by * and **.
```

This is implemented by https://github.com/wwkimball/yamlpath/commit/a782af67dbc5cbf2e855d85a0f17a57fa96733bd and I've similarly [updated the documentation](https://github.com/wwkimball/yamlpath/wiki/Python-Library) to read:


> Be aware that the `get_nodes` method will attempt to create any missing document structure in order to match the supplied YAML Path unless you set `mustexist=True`.  The default is `False`, which is only to ensure parity with the default behavior of the `set_value` method.  Be careful to either set `mustexist=True` or provide a default value when deliberately using this.

**Comment by wwkimball** [comment]
I don't know whether anyone but me can see the Project board on the repo, but I've added this issue to the board for the upcoming 3.7.0 release.  Unless anyone strongly feels otherwise, I'm going to call this issue closed with [the proposal above](https://github.com/wwkimball/yamlpath/issues/206#issuecomment-1376499672).

I'd like to flip the default behavior of the get_nodes method to specifically require `mustexist=False` in order to auto-generate missing DOM, but because it's a breaking behavioral change, I'm going to shelve that for a 4.x release.  I'm happy to hear arguments for an earlier change.

This issue will automatically close when I've published the next 3.x release, which I hope to do within the week.
